src/api/HttpApi.js

```diff
@@ -0,0 +1,269 @@
+    this._externalOrigin = externalOrigin;
+    this._apiBase = apiBase;
+
+    this._loader = new DataLoader(paths =>
+      Promise.all(
+        paths.map(path =>
+          // Don't fail the entire batch on a single failed request.
+          this._get(path).catch(e => e),
+        ),
+      ),
+    );
+  }
+
+  getResponseData({ data, meta }: { data?: mixed, meta?: mixed }) {
+    const translatedData: any = data;
+    if (meta) translatedData.meta = meta;
+    return translatedData;
+  }
+
+  getHeaders() {
+    return {
+      Authorization: this.authorization,
+    };
+  }
+
+  async get<T>(path: string, args?: Args): Promise<?T> {
+    const item = await this._loader.load(this.makePath(path, args));
+    if (item instanceof Error) {
+      throw item;
+    }
+
+    return item;
+  }
+
+  async getPaginatedConnection(
+    path: string,
+    { after, first, ...args }: Args,
+  ): Promise<mixed> {
+    const items = await this.get(
+      this.makePath(path, {
+        ...args,
+        cursor: after,
+        limit: first,
+        pageSize: first,
+      }),
+    );
+
+    if (!items) {
+      return null;
+    }
+
+    if (items.meta) {
+      const { cursors, hasNextPage } = items.meta;
+
+      // These connections only paginate forward, so the existence of a prev
+      // page doesn't make any difference, but this is the correct value.
+      const hasPreviousPage = !!after;
+
+      return {
+        edges: items.map((item, i) => ({
+          node: item,
+          cursor: cursors[i],
+        })),
+        pageInfo: {
+          hasNextPage,
+          endCursor: cursors[cursors.length - 1],
+          hasPreviousPage,
+        },
+      };
+    }
+    throw new Error('unexpected items format');
+  }
+
+  async getUnpaginatedConnection(path: string, args: Args): Promise<mixed> {
+    const apiArgs = omit(args, PAGINATION_ARG_KEYS);
+    const paginationArgs = pick(args, PAGINATION_ARG_KEYS);
+
+    // XXX Need to cast the result of the get to a list
+    const items = await this.get(this.makePath(path, apiArgs));
+    if (!Array.isArray(items)) {
+      throw new Error('Runtime Casting Exception');
+    }
+    return connectionFromArray(items, paginationArgs);
+  }
+
+  async getValidationResult(
+    path: string,
+    args: Args,
+  ): Promise<ValidationResult> {
+    try {
+      await this.get(path, args);
+    } catch (e) {
+      if (e instanceof HttpError && (e.status === 422 || e.status === 409)) {
+        return {
+          valid: false,
+          errors: e.errors,
+        };
+      }
+
+      throw e;
+    }
+
+    return {
+      valid: true,
+      errors: null,
+    };
+  }
+
+  async fetch<T>(method: string, path: string, data?: Data): Promise<?T> {
+    const json = await request({
+      method,
+      url: this._getUrl(path, false),
+      headers: this.getHeaders(),
+      data: (translateKeys(data, this.serializeKey): any),
+      files: this.request.files.map(({ fieldname, ...file }) => ({
```

Comment:
might need some check here, since `this.request.files` will probably be falsy if multer isn't around

Reply:
Bah, yeah it's all annoyingly leaky :/

Reply:
I would just make this bit an abstract method and have the application server implement it. I think it's a workable way to handle the key translation for now, too.